### ISSUE TITLE
Teleport should drop running as an unprivileged user if it isn't running a 'node' process

### ISSUE THREAD

**Issue by webvictim** [gravitational/teleport#2759]
As discussed on Slack.

Currently, Teleport will happily run as root even when the config file or CLI parameters don't specify that a `node` process needs to run. `auth` and `proxy` will run fine as a non-root user, and it'd be a nice security feature to warn the user that root privileges aren't needed when `node` isn't enabled and drop to an unprivileged user.

This would also be useful inside a Docker container and help us to comply with security best practices.

**Comment by webvictim** [comment]
cc @russjones

**Comment by kontsevoy** [comment]
@webvictim but why would that be a Teleport ticket? 

"Back in the day" daemons used to have a configuration setting (like `libvirtd` or `postfix`) called `user` to indicate _drop my privileges to "user"_. But since `systemd` (or upstart before it) took over, this functionality has moved to the supervisor, basically users can already create a system user and set it to `teleport.service` unit file.

Why would we need to duplicate this functionality inside teleport?

**Comment by webvictim** [comment]
@kontsevoy I agree in principle, but people don't always use systemd/upstart to run processes - we're using `dumb-init` within our own Docker containers for example which doesn't have this functionality.

The reason this ticket was created is because we always build our own official Docker containers with Teleport running as `root`  - this is so that if the teleport daemon is running a node process and *does* need root privileges (to spawn sessions as any given user) it will have them. In many cases, though, teleport isn't running a node process at all and so it running as `root` is just an unnecessary security risk. This also makes us fall foul of numerous Kubernetes and Docker recommendations that containers should never be run as `root` unless absolutely necessary.

I figured that rather than trying to do some horrific config parsing within our container scripts to determine from the config whether we need to run as `root` or not, it would be a nicer feature to just do a simple check within the Teleport code like `if (euid == 0 && nodeProcess == notRunning) { log.Warn("The teleport process has been started with root privileges but does not need them") }`, perhaps along with willingly dropping its own privileges to an unprivileged user (the big question, of course, would be _which_ user it should drop to if one isn't already configured)

**Comment by webvictim** [comment]
Yes, but we're not providing any examples of how to do that within our own Helm charts or recommended deployments. Out of the box, our containers' configuration goes against security best practices (which is pretty poor for a company that wants to help others improve security)

**Comment by klizhentas** [comment]
> within our own Helm charts or recommended deployments.

Our terraform / cloudformation do not run proxies and auth servers as root. 

Our helm charts could be probably improved, I agree with this part.

**Comment by webvictim** [comment]
I've updated the scope of this ticket to adding a warning to Teleport when running as `root` if root privileges aren't needed.

I also created https://github.com/gravitational/teleport/issues/2909 to improve the security model used in our Helm and Docker examples.

**Comment by stmuraka** [comment]
I'd also like to see the Teleport containers to run as non-`root`. i.e. user `teleport`
we're overriding the root user by setting a kube Security Context. it would be nice that it didn't have elevated privileges by default.

@benarent
